core/payment/src/processor.rs

```diff
@@ -306,8 +311,117 @@ impl DriverRegistry {
 const DB_LOCK_TIMEOUT: Duration = Duration::from_secs(30);
 const REGISTRY_LOCK_TIMEOUT: Duration = Duration::from_secs(30);
 
+struct TimedMutex {
+    mutex: Mutex<DbExecutor>,
+    sender: Option<UnboundedSender<TimedMutexTaskMessage>>,
+    counter_task: Option<JoinHandle<()>>,
+}
+
+use tokio::sync::MutexGuard;
+
+enum TimedMutexTaskMessage {
+    Start(String),
+    Finish,
+}
+
+struct TimedMutexGuard<'a> {
+    mutex_guard: MutexGuard<'a, DbExecutor>,
+    sender: &'a Option<UnboundedSender<TimedMutexTaskMessage>>,
+}
+
+impl Drop for TimedMutexGuard<'_> {
+    fn drop(&mut self) {
+        if let Some(sender) = &self.sender {
+            if let Err(e) = sender.send(TimedMutexTaskMessage::Finish) {
+                log::error!("Cannot send fininsh to counter task {e}");
+            }
+        }
+    }
+}
+
+impl<'a> Deref for TimedMutexGuard<'a> {
+    type Target = MutexGuard<'a, DbExecutor>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.mutex_guard
+    }
+}
+
+impl TimedMutex {
+    fn new(db: DbExecutor) -> Self {
+        let (sender, mut receiver) =
+            tokio::sync::mpsc::unbounded_channel::<TimedMutexTaskMessage>();
+
+        let counter_task = tokio::spawn(async move {
+            loop {
+                // wait for start or close without timeout
+                let task_name = match receiver.recv().await {
+                    None => return,
+                    Some(TimedMutexTaskMessage::Start(x)) => x,
+                    Some(TimedMutexTaskMessage::Finish) => {
+                        panic!("Unexpected finish")
+                    }
+                };
+
+                let mut counter = 0;
+                loop {
+                    match tokio::time::timeout(Duration::from_secs(10), receiver.recv()).await {
+                        Err(_) => {
+                            log::error!("Long running task: {task_name}!");
+                            counter += 1;
+                            // five minutes
+                            if counter > 30 {
+                                exit(41);
+                            }
+                        }
+                        Ok(None) => panic!("Unexpected mpsc close."),
+                        Ok(Some(TimedMutexTaskMessage::Finish)) => break,
+                        Ok(Some(TimedMutexTaskMessage::Start(_))) => panic!("Unexpected start"),
+                    }
+                }
+            }
+        });
+
+        Self {
+            mutex: Mutex::new(db),
+            sender: Some(sender),
+            counter_task: Some(counter_task),
+        }
+    }
+
+    async fn timeout_lock(
+        &self,
+        duration: Duration,
+        name: &str,
+    ) -> Result<TimedMutexGuard<'_>, Elapsed> {
+        let result = tokio::time::timeout(duration, self.mutex.lock())
+            .await
+            .map_err(|e| {
+                log::info!("Failed to lock mutex in scenario {0}", name);
```

Comment:
Please ensure that we will collect these logs (I don't know what production log level setting is).
edit: I looked quickly at [cloudwatch insights](https://eu-central-1.console.aws.amazon.com/cloudwatch/home?region=eu-central-1#logsV2:logs-insights$3FqueryDetail$3D~(end~0~start~-86400~timeType~'RELATIVE~tz~'UTC~unit~'seconds~editorString~'fields*20*40timestamp*2c*20*40message*2c*20*40logStream*2c*20*40log*0a*7c*20filter*20*40message*20like*20*22INFO*22*20and*20*40message*20like*20*22payment*22*20and*20*40message*20like*20*22processor*22*0a*7c*20sort*20*40timestamp*20desc*0a*7c*20limit*201000~queryId~'f8f0af25-fda1-4b67-9a22-8fab8ef2107d~source~(~'*2fecs*2fscalepoint-prod-yagna-gh-log-group~'*2fecs*2fscalepoint-prod-yagna-vm-log-group))) and it seems like we do.